### DISCUSSION TITLE
Externalizing Inline script data to reduce HTML file size for improved SEO indexing

### DISCUSSION THREAD

**Discussion by RykGitProjects** [sveltejs/kit#12491] [Ideas]
Hello!

I am an enthusiastic user of SvelteKit for web development, appreciating its speed and ease of use. To optimize performance, I've implemented various strategies such as loading libraries only after user scrolls and loading Google Tag Manager on user interaction. These enhancements cater to both user experience and search engine requirements since bots typically do not simulate interactions.

Issue:
However, a significant challenge I face relates to SEO with Bing Search, which has indicated that many pages exceed the 125 kB size limit for the HTML documents. This is primarily due to the large inline scripts that SvelteKit appends for hydration and SSR, embedded directly within the main HTML:

<script>
  {
    __sveltekit_erhfef = {
      base: new URL("../..", location).pathname slice(0, -1)
    };

    const element = document.currentScript.parentElement;

    const data = [<data fetched by +page.server.js>];

    Promise.all([
      import("../../_app/immutable/entry/start.B791bc_E.js"),
      import("../../_app/immutable/entry/app.CIH1Uats.js")
    ]).then(([kit, app]) => {
      kit.start(app, element, {
        node_ids: [0, 10],
        data,
        form: null,
        error: null
      });
    });
  }
</script>

In my specific case, the inclusion of this script increases the source code of a page from about 80 kB without the script to approximately 190 kB with it. This nearly triples the file size, significantly impacting indexing as Bing does not prefer to store large HTML files.

Suggestion:
Would it be feasible to introduce a feature in SvelteKit that allows for the externalization of this script data, akin to linking CSS files externally? This modification would considerably reduce the size of the initial HTML document, which is crucial for meeting Bing's indexing requirements.

While further data optimization is possible, such as compressing or restructuring the data, the fundamental issue of large HTML document size remains. An option to externalize the script data would directly address this challenge.

Thank you for considering this feature request. Any feedback or suggestions from the community would be greatly appreciated!

**Comment by benmccann** [comment]
Could you share the URL of the site you're talking about?

If the SvelteKit startup script is taking 110 kB it would seem that you have a very large number of scripts that it is loading. Perhaps you could use a custom Vite chunking to reduce the number of output scripts that need to be loaded by the inline script: https://v3.vitejs.dev/guide/build.html#chunking-strategy

**Reply by Rich-Harris** [reply]
> it would seem that you have a very large number of scripts that it is loading

it's more likely to be inlined data

**Reply by RykGitProjects** [reply]
Richard is right, it is about amount of data
Link: bizraport.pl/krs/0000140562/cedrob-spolka-akcyjna

**Reply by RykGitProjects** [reply]
@benmccann - do you see any possibility to decrease page size, which bing saves (probably without js) ?

**Reply by benmccann** [reply]
You could try dynamic rendering where you strip that block out if it is the bing bot user agent

**Reply by RykGitProjects** [reply]
Seems complicated, as I would need to decide if it is bing bot user agent in some +page.server.js, as I won't be able to run any js on front (js is inactive when bot uses page). Still thx for letting me know. Seems like not a common problem.

**Reply by benmccann** [reply]
You would do it in `handle` rather than a `+page.server.js`. The fact that the bot basically ignores the script is what makes it possible to strip it out

**Reply by RykGitProjects** [reply]
Thank you, I will work on it
